src/odf/dataset-impl/src/entities/metadata_chain_validators.rs

```diff
@@ -856,6 +856,47 @@ impl<'a> ValidateSetDataSchemaVisitor<'a> {
             vocab: None,
         })
     }
+
+    fn check_allowed_schema_migration(diff: &DataSchemaDiff<'_>) -> bool {
+        diff.items
+            .iter()
+            .all(Self::check_allowed_schema_migration_rec)
+    }
+
+    fn check_allowed_schema_migration_rec(diff: &DataSchemaDiffItem<'_>) -> bool {
+        #[expect(clippy::match_same_arms)]
+        match diff {
+            DataSchemaDiffItem::SchemaAttributesChanged => true,
+            DataSchemaDiffItem::FieldAttributesChanged => true,
+            DataSchemaDiffItem::FieldAdded { field_rhs } => {
+                if field_rhs.is_optional() {
+                    // It's OK to introduce a new optional field as this maintains compatibility
+                    // from publisher and consumer perspective
+                    true
```

Comment:
I understand consumer's view, but not publisher's.
I.e., what do we do, if we attempt to use v2 API to write to non-migrated data room?
I could imagine that publisher pushes new optional fields to a dataset with an old schema, and they will be just ignored, but the publisher still needs to find out of incompatibility somehow

Reply:
> I.e., what do we do, if we attempt to use v2 API to write to non-migrated data room?

This situation should not happen in `prod` - we will run migrations before re-enabling access token.

But in a general case I think we need a more expressive ingest API that allows to:
- Control whether to return an error if input data has fields that are not present in the schema, or to ignore them
- Assert that schema is exactly the same as what writer expects (e.g. `ODF-If-Schema-Hash-Matches: f16..c77`) aborting ingest if it isn't